pkg/apis/network/v1/egressrouter_types.go

```diff
@@ -0,0 +1,121 @@
+// +kubebuilder:resource:shortName=egr
+// +kubebuilder:printcolumn:name="EgressRouters",type=string,JSONPath=".status.egressRouter"
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=".status.status"
+type EgressRouter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec EgressRouterSpec `json:"spec"`
+
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status EgressRouterStatus `json:"status,omitempty"`
+}
+
+// EgressRouterSpec contains the configuration for an egress router.
+// +k8s:openapi-gen=true
+// +kubebuilder:validation:Required
+type EgressRouterSpec struct {
+	// Specification of interface to create/use. e.g; macvlan
+	// +kubebuilder:validation:Required
+	NetworkInterface EgressRouterInterface `json:"networkInterface"`
+
+	// IP configuration arguments.
+	// +kubebuilder:validation:Required
+	IP IPConfig `json:"ip"`
+}
+
+// EgressRouterInterface contains the configuration of interface to create/use.
+// +kubebuilder:validation:Required
+type EgressRouterInterface struct {
+	// Arguments specific to the interfaceType macvlan
+	// +kubebuilder:validation:Required
+	MacVLAN VLANConfig `json:"macVLAN"`
+}
+
+// VLANConfig consists of arguments specific to the EgressRouterInterfaceType
+// +kubebuilder:validation:Required
+type VLANConfig struct {
+	// Usually used for macvlan interface.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Mode string `json:"mode,omitempty"`
+
+	// Name of the master inteface. Usually used for macvlan and ipvlan. Need not be specified if it can be inferred from the IP address.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Master string `json:"master,omitempty"`
+}
+
+// IPConfig contains the configuration for the ip arguments
+// +kubebuilder:validation:Required
+type IPConfig struct {
+	// List of IP addresses to configure on the interface.
+	// +kubebuilder:validation:Required
+	Addresses []EgressRouterAddress `json:"addresses"`
+
+	// List of CIDR blocks that the pod is allowed to connect to via this interface. If not provided, the pod can connect to any destination.
+	// +kubebuilder:validation:Optional
+	// +optional
+	AllowedDestinations []CIDR `json:"allowedDestinations,omitempty"`
+}
+
+// EgressRouterAddress contains a pair of ip cidr and gateway to be configured on the router's interface
+// +kubebuilder:validation:Required
+type EgressRouterAddress struct {
+	// IP is the address to configure on the router's interface. Can be IPv4 or IPv6.
+	// +kubebuilder:validation:Required
+	IP CIDR `json:"ip"`
+	// IP address of the next-hop gateway, if it cannot be automatically determined. Can be IPv4 or IPv6.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Gateway string `json:"gateway,omitempty"`
+}
+
+// CIDR represents an IPv4 or IPV6 CIDR block.
```

Comment:
Probably want to add a simple pattern matcher here - it won't be perfect (a proper one is like 4kb) but something simple would be good. It should be dual-stack :-).

Reply:
yea I removed the pattern because the actual pattern for ipv6 was too huge, instead daniel has agreed to manually add the check in the admission controller.